### ISSUE TITLE
Descriptors shouldn't be considered callable inside class definition

### ISSUE THREAD

**Issue by refi64** [python/mypy#2267]
The following is an error in Python but passes mypy's checks:

``` python
class X:
    @staticmethod
    def f(): pass
    f()
```

Error:

```
Traceback (most recent call last):
  File "cls.py", line 1, in <module>
    class X:
  File "cls.py", line 4, in X
    f()
TypeError: 'staticmethod' object is not callable
```

Same thing goes for `classmethod`.

**Comment by gvanrossum** [comment]
Ugh. Mypy's lack of understanding of class vs. instance attributes strikes again. :-(

**Comment by elazarg** [comment]
I would expect mypy to mimic the runtime - treat the current block as if it was not inside a class, then send the definitions to a "metaclass" (hard coded, in the case to `type`) to handle. It will also yield a natural extension point for a plugin: implement your home-made "metaclass".

Does that make sense at all? As far as I understand, this isn't how it's done, currently - not even remotely similar.

**Comment by gvanrossum** [comment]
Would be good to fix but probably complicated.

**Comment by gvanrossum** [comment]
I *think* you meant "decorators" in the title, not "descriptors", right? (A decorator uses `@` syntax -- a descriptor is something that implements `__get__` and (optionally) `__set__` and `__delete__`.

**Comment by kirbyfan64** [comment]
@gvanrossum Well, I meant descriptors because of the `staticmethod` descriptor.

**Comment by gvanrossum** [comment]
Ah, got it.

**Comment by erictraut** [comment]
This doesn't appear to generate a runtime exception with newer versions of Python, so it's no longer a false negative in mypy.

**Comment by JelleZijlstra** [comment]
The runtime change is from https://bugs.python.org/issue43682 which was included in Python 3.10.

**Comment by hauntsaninja** [comment]
Closing as won't fix, though if anyone feels inspired to write a PR for 3.9 feel free to
